src/netCDF4/_netCDF4.pyx

```diff
@@ -3333,9 +3336,10 @@ of existing (sub-) groups and their variables.
             for group in groups:
                 group.set_ncstring_attrs(value) # recurse into subgroups...
 
+    @functools.lru_cache(maxsize=128)
```

Comment:
Adding the decorator here makes things simpler but can lead to memory leaks, see https://github.com/python/cpython/issues/64058. However, b/c I could not make the [most common fix](https://www.youtube.com/watch?v=sVjtp6tGo0g) for this work here and I believe with the cache clear in the `__dealloc__` we cover the cases of opening and closing multiple datasets, I don't think this would be a problem. (I may be wrong here! I'm not an expert and I'm envisioning a subset of the use cases I know.)

What do you think @jswhit?

The alternative would be to factor this function out into a private one that takes `self.variables` and `self._vars`, cache that one, and use it in the class.

Reply:
@ocefpaf  I'm OK with this approach unless there are significant memory leaks.   I guess you could create a test for memory leaks by repeatedly opening and closing the same file and invoking get_variables_by_attributes multiple times to see what happens to memory usage.